File: 0383-ransom-note/0383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const magazinLetters = new Map();
    
    for(i of magazine) {
        const currentCount = magazinLetters.get(i) ?? 0
        magazinLetters.set(i, currentCount+1)
    }

    for(i of ransomNote) {
        const currentCount = magazinLetters.get(i) ?? 0
        if(!currentCount) return false;
        
        magazinLetters.set(i, currentCount-1)
    }

    return true;
};

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_letters = {}

        for char in magazine:
            magazine_letters[char] = magazine_letters.get(char, 0) + 1
            # get(char, 기본값) ~> char 키가 존재하지 않으면 기본값을 반환한다.

        for char in ransomNote:
            if char not in magazine_letters or magazine_letters[char] == 0:
                return False
            
            magazine_letters[char] -= 1
        
        return True